File: bindings/java/src/test/java/org/indunet/robot/bus/InprocContextTest.java
package org.indunet.robot.bus;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;
import org.junit.jupiter.api.Test;

/** Same-process inproc requires a shared {@link Context} with the embedded broker. */
class InprocContextTest {
    @Test
    void inprocPubSubWithSharedContext() throws Exception {
        try (Context ctx = new Context();
                Broker broker = new Broker(ctx, inprocBrokerOptions());
                Node sub = Node.inproc(ctx, "inproc-sub");
                Node pub = Node.inproc(ctx, "inproc-pub")) {
            Thread.sleep(150);

            AtomicInteger hits = new AtomicInteger();
            sub.createSubscription(
                    "/inproc/demo",
                    (payload) -> {
                        assertEquals("hello-inproc", new String(payload, StandardCharsets.UTF_8));
                        hits.incrementAndGet();
                    });
            sub.start();
            Thread.sleep(100);

            try (TopicPublisher topic = pub.createPublisher("/inproc/demo")) {
                long deadline = System.currentTimeMillis() + 5000;
                while (hits.get() < 1 && System.currentTimeMillis() < deadline) {
                    topic.publish("hello-inproc".getBytes(StandardCharsets.UTF_8));
                    Thread.sleep(20);
                }
            }

            assertTrue(hits.get() >= 1, "expected at least one inproc message");
            sub.shutdown();
            sub.waitForShutdown();
        }
    }

    @Test
    void inprocActionGoalHandle() throws Exception {
        try (Context ctx = new Context();
                Broker broker = new Broker(ctx, inprocBrokerOptions());
                Node server = Node.inproc(ctx, "inproc-action-server");
                Node clientNode = Node.inproc(ctx, "inproc-action-client")) {
            Thread.sleep(150);

            server.createActionServer(
                    "/inproc/action",
                    body ->
                            List.of(
                                    new ActionPhase(
                                            "FEEDBACK",
                                            ("step:" + new String(body, StandardCharsets.UTF_8))
                                                    .getBytes(StandardCharsets.UTF_8)),
                                    new ActionPhase(
                                            "RESULT",
                                            ("done:" + new String(body, StandardCharsets.UTF_8))
                                                    .getBytes(StandardCharsets.UTF_8))));
            server.start();
            Thread.sleep(100);

            List<String> feedback = new ArrayList<>();
            try (ActionClient action = clientNode.createActionClient("/inproc/action");
                    ActionGoalHandle goal =
                            action.sendGoal(
                                    "move".getBytes(StandardCharsets.UTF_8),
                                    null,
                                    3.0,
                                    message ->
                                            feedback.add(
                                                    new String(
                                                            message.getBody(),
                                                            StandardCharsets.UTF_8)))) {
                assertEquals("/inproc/action", goal.actionName());
                assertFalse(goal.goalId().isEmpty());
                assertEquals(
                        "done:move",
                        new String(goal.result(3.0).getBody(), StandardCharsets.UTF_8));
                assertEquals(List.of("step:move"), feedback);
            }

            server.shutdown();
            server.waitForShutdown();
        }
    }

    /** Ephemeral TCP binds (`:0`), but keep inproc (tcpOnly=false). */
    private static BrokerOptions inprocBrokerOptions() {
        return new BrokerOptions(
                "tcp://127.0.0.1:0",
                "tcp://127.0.0.1:0",
                "tcp://127.0.0.1:0",
                "tcp://127.0.0.1:0",
                "tcp://127.0.0.1:0",
                "tcp://127.0.0.1:0",
                "127.0.0.1:0",
                null,
                false,
                true);
    }
}

File: bindings/android/src/test/kotlin/org/indunet/robot/bus/InprocContextTest.kt
package org.indunet.robot.bus

import java.io.IOException
import java.net.ServerSocket
import java.nio.charset.StandardCharsets
import java.util.concurrent.atomic.AtomicInteger
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

/** Same-process inproc requires a shared [Context] with the embedded broker. */
class InprocContextTest {
    @Test
    fun inprocPubSubWithSharedContext() {
        Context().use { ctx ->
            Broker(ctx, inprocBrokerOptions()).use { _ ->
                Node.inproc(ctx, "inproc-sub").use { sub ->
                    Node.inproc(ctx, "inproc-pub").use { pub ->
                        Thread.sleep(150)

                        val hits = AtomicInteger()
                        sub.createSubscription(
                            "/inproc/demo",
                            { _, payload ->
                                assertEquals(
                                    "hello-inproc",
                                    String(payload, StandardCharsets.UTF_8),
                                )
                                hits.incrementAndGet()
                            },
                        )
                        sub.start()
                        Thread.sleep(100)

                        pub.createPublisher("/inproc/demo").use { topic ->
                            val deadline = System.currentTimeMillis() + 5000
                            while (hits.get() < 1 && System.currentTimeMillis() < deadline) {
                                topic.publish("hello-inproc".toByteArray(StandardCharsets.UTF_8))
                                Thread.sleep(20)
                            }
                        }

                        assertTrue("expected at least one inproc message", hits.get() >= 1)
                        sub.shutdown()
                        sub.waitForShutdown()
                    }
                }
            }
        }
    }

    @Test
    fun inprocActionGoalHandle() {
        Context().use { ctx ->
            Broker(ctx, inprocBrokerOptions()).use { _ ->
                Node.inproc(ctx, "inproc-action-server").use { server ->
                    Node.inproc(ctx, "inproc-action-client").use { clientNode ->
                        Thread.sleep(150)

                        server.createActionServer(
                            "/inproc/action",
                            ActionHandler { body ->
                                val payload = String(body, StandardCharsets.UTF_8)
                                listOf(
                                    ActionPhase(
                                        "FEEDBACK",
                                        "step:$payload".toByteArray(StandardCharsets.UTF_8),
                                    ),
                                    ActionPhase(
                                        "RESULT",
                                        "done:$payload".toByteArray(StandardCharsets.UTF_8),
                                    ),
                                )
                            },
                        )
                        server.start()
                        Thread.sleep(100)

                        val feedback = mutableListOf<String>()
                        clientNode.createActionClient("/inproc/action").use { action ->
                            action.sendGoal(
                                "move".toByteArray(StandardCharsets.UTF_8),
                                null,
                                3.0,
                            ) { message ->
                                feedback.add(String(message.body, StandardCharsets.UTF_8))
                            }.use { goal ->
                                assertEquals("/inproc/action", goal.actionName())
                                assertTrue(goal.goalId().isNotEmpty())
                                assertEquals(
                                    "done:move",
                                    String(goal.result(3.0).body, StandardCharsets.UTF_8),
                                )
                                assertEquals(listOf("step:move"), feedback)
                            }
                        }

                        server.shutdown()
                        server.waitForShutdown()
                    }
                }
            }
        }
    }

    /** Ephemeral TCP binds, but keep inproc (tcpOnly=false). */
    @Throws(IOException::class)
    private fun inprocBrokerOptions(): BrokerOptions =
        BrokerOptions(
            messageXsubBind = "tcp://127.0.0.1:${freePort()}",
            messageXpubBind = "tcp://127.0.0.1:${freePort()}",
            serviceFrontendBind = "tcp://127.0.0.1:${freePort()}",
            serviceBackendBind = "tcp://127.0.0.1:${freePort()}",
            actionFrontendBind = "tcp://127.0.0.1:${freePort()}",
            actionBackendBind = "tcp://127.0.0.1:${freePort()}",
            apiListen = "127.0.0.1:${freePort()}",
            consoleListen = null,
            tcpOnly = false,
            noConsole = true,
        )

    @Throws(IOException::class)
    private fun freePort(): Int =
        ServerSocket(0).use { socket ->
            socket.reuseAddress = true
            socket.localPort
        }
}
